### PR TITLE
apply PREFER_DAY_OF_MONTH in date.py

### PR DESCRIPTION
**Pull request by noviluni**
fixes https://github.com/scrapinghub/dateparser/issues/607

### CODE REVIEW

**Comment by noviluni** [comment]
(Explanation and example in a comment in the PR)

**Comment by Gallaecio** [comment]
What I found most strange about #607 was that, depending on whether the date format input parameter was a list (`['%m/%Y']`) or a string (`'%m/%Y'`), the output date was different.

I think that is something we need to address and cover in the tests before we can consider #607 solved.

**Comment by noviluni** [comment]
@Gallaecio 
I think there was a confusion.

Currently, doing this:
```python
dateparser.parse('02/2013', ['%m/%Y'])                                                                                                                                                                  
>>> datetime.datetime(2013, 2, 28, 0, 0)
```
Is the same than doing this:
```python
>>> dateparser.parse('02/2013', '%m/%Y')                                                                                                                                                                    
/home/marc/workprojects/dateparser/dateparser/date.py:142: FutureWarning: Date formats should be list, tuple or set of strings
  warn(_DateLocaleParser.DATE_FORMATS_ERROR_MESSAGE, FutureWarning)
datetime.datetime(2013, 2, 28, 0, 0)
```
(except for the deprecation warning). In the example in the issue, the second statement is
```
datetime.strptime('02/2013', '%m/%Y')
``` 
which is using the `datetime.strptime`, not the `dateparse.parse` method.

**Comment by Gallaecio** [comment]
:man_facepalming:

File: dateparser/utils/__init__.py [hunk 1]
```diff
@@ -133,6 +135,24 @@ def apply_timezone_from_settings(date_obj, settings):
     return date_obj
 
 
+def get_last_day_of_month(year, month):
+    return calendar.monthrange(year, month)[1]
+
+
+def set_correct_day_from_settings(date_obj, settings, last_day=None, current_day=None):
```
Comment (Gallaecio) [inline dateparser/utils/__init__.py]: :lipstick: What about `apply_day_of_the_month_from_settings`, to be consistent with `apply_timezone_from_settings`?

File: dateparser/parser.py [hunk 2]
```diff
@@ -430,16 +431,11 @@ def _correct_for_day(self, dateobj):
             return dateobj
 
         _, tail = calendar.monthrange(dateobj.year, dateobj.month)
```
Comment (noviluni) [inline dateparser/parser.py]: I have just seen that we don't really need to generate here the `tail` as can be directly generated in the `set_correct_day_from_settings`. I will change it when I have a moment.

File: dateparser/utils/__init__.py [hunk 3]
```diff
@@ -133,6 +135,24 @@ def apply_timezone_from_settings(date_obj, settings):
...
+    """ Set correct day attending the `PREFER_DAY_OF_MONTH` setting."""
```
Comment (noviluni) [inline dateparser/utils/__init__.py]: any better idea for the naming of this function?

Comment (Gallaecio) [reply dateparser/utils/__init__.py]: https://github.com/scrapinghub/dateparser/pull/611#discussion_r368165205

**Comment by noviluni** [comment]
Note to self: I could also add a unit test for the function `get_last_day_of_month` (including all months and at least one example of leap year).

File: dateparser/utils/__init__.py (hunk 1 above)
Comment (noviluni) [reply dateparser/utils/__init__.py]: Hmmm... This was my first approach, but then I didn't like that naming. First of all "apply" seems to mean that you are "adding" something. It maybe has sense for the timezone, as sometimes you add it, but I'm not sure if this has sense for the day as what you are doing is to change it.

Secondly, `apply_day_of_the_month_from_settings` seems to mean that it is getting the day from the settings; but what it is doing is using the rules in the settings to find the correct day.

For both reasons I prefer this naming: `set_correct_day_from_settings`.

What do you think @Gallaecio ? Any better idea?

**Pull request merged**